design/0001-metallb-crd.md

```diff
@@ -0,0 +1,173 @@
+# MetalLB Custom Resource Controller 
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Testing](#testing)
+- [Debugging and Troubleshooting](#debugging-and-troubleshooting)
+<!-- /toc -->
+
+## Summary
+Today MetalLB allows only configuration of layer2 and BGP protocols using Configmap objects, 
+which is simple from configuration perspective and doesn't require cluster-admin rights,
+however it doesn't have validation on the API server level, no support for API versioning and 
+no fine-grained permission model via RBAC.
+
+## Motivation
+Custom resource definition "CRD" takes advantage of kubernetes extensibility to deliver the automation advantages of cloud services like 
+provisioning, scaling and backup/restore.
+
+### Goals
+Configuring MetalLB layer2 and BGP using kubernetes custom resource "CRD" instead of ConfigMap.
```

Comment:
Are there any other goals/non-goals? 
* Do we have any goals regarding stability of the CRD?
* I presume that we intend to expose all current ConfigMap configuration via the CRD?

Reply:
using k8s CRD enable us to support versioning to help with version upgrades , so that support is built in CRD using Conversion web hook.